File: src/main/java/masonry/transition/TransitionBase.java
package masonry.transition;

import masonry.api.Transition;
import net.scriptgate.common.Point;

public abstract class TransitionBase implements Transition {

    public static float TRANSITION_SPEED = 1000;
    private float transition = 0;
    private boolean isCompleted = false;

    public Point getLocationAt(double elapsedTime) {
        if (isCompleted) {
            return new Point(toX(), toY());
        }

        if (transition + elapsedTime > TRANSITION_SPEED) {
            isCompleted = true;
            return new Point(toX(), toY());
        } else {
            transition += elapsedTime;
            return getLocation();
        }
    }

    public abstract Point getLocation();

    @Override
    public float getPercentage() {
        return transition / TRANSITION_SPEED;
    }

    @Override
    public boolean isCompleted() {
        return isCompleted;
    }
}

File: src/main/kotlin/net/scriptgate/masonry/transition/TransitionBase.kt
package net.scriptgate.masonry.transition

import net.scriptgate.common.Point
import net.scriptgate.masonry.api.Transition

abstract class TransitionBase : Transition {
    private var transition = 0f

    private var transitionDone : Boolean = false

    override var isCompleted: Boolean
        get() = transitionDone
        set(value) {transitionDone = value}


    abstract val location: Point

    override val percentage: Float
        get() = transition / TRANSITION_SPEED

    override fun getLocationAt(elapsedTime: Double): Point {
        if (isCompleted) {
            return Point(toX(), toY())
        }

        if (transition + elapsedTime > TRANSITION_SPEED) {
            isCompleted = true
            return Point(toX(), toY())
        } else {
            transition += elapsedTime.toFloat()
            return location
        }
    }

    companion object {

        var TRANSITION_SPEED = 1000f
    }
}
